### PR TITLE
feat: add cloudspanner functionality.

### PR DESCRIPTION
**Pull request by rsyi**
Closes https://github.com/dataframehq/whale/issues/68.

### CODE REVIEW

File: pipelines/whale/extractor/cloud_spanner_extractor.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        if not self._extract_iter:
+            self._extract_iter = self._get_extract_iter()
+        try:
+            return next(self._extract_iter)
+        except StopIteration:
+            return None
+
+    def get_scope(self):
+        # type: () -> str
+        return 'extractor.presto_table_metadata'
```
Comment (manuelzander) [inline pipelines/whale/extractor/cloud_spanner_extractor.py:91]: @rsyi Typo - `return 'extractor.cloudspanner_table_metadata'`? 😊

File: pipelines/whale/extractor/cloud_spanner_extractor.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+            for _, group in groupby(headered_results, self._get_table_key):
+                columns = []
+
+                for row in group:
+                    last_row = row
+                    columns.append(ColumnMetadata(row['col_name'], None,
+                                                  row['col_type'], row['col_sort_order']))
+
+            yield TableMetadata(self._instance_id,
```
Comment (manuelzander) [inline pipelines/whale/extractor/cloud_spanner_extractor.py:113]: @rsyi Ran it locally, I think `yield...` is missing one indent (yield for each table, not only once).

Comment (rsyi) [reply pipelines/whale/extractor/cloud_spanner_extractor.py:113]: amazing catch, thanks @manuelzander!

File: pipelines/whale/extractor/cloud_spanner_extractor.py (hunk 1 above)
Comment (rsyi) [reply pipelines/whale/extractor/cloud_spanner_extractor.py:91]: Great catch.

**Comment by manuelzander** [comment]
FYI - @biomunky and me have been hacking around with a Spanner extractor: https://github.com/dataframehq/whale/compare/master...manuelzander:master

**Comment by manuelzander** [comment]
Another general point: when I ran this locally I didn't seem to get any data types (all `None`) - maybe worth raising an issue?

**Comment by rsyi** [comment]
@manuelzander LOVE IT. Let's go with yours - it's much more robust. I don't use spanner, so I didn't even realize they had an orm over the metadata (hence me missing all the types).

**Comment by manuelzander** [comment]
> @manuelzander LOVE IT. Let's go with yours - it's much more robust. I don't use spanner, so I didn't even realize they had an orm over the metadata (hence me missing all the types).

Mine does extract data types BUT this is using https://github.com/shinichi-takii/ddlparse and `ddlparse` currently has limited support for ARRAY or STRING data types, etc. (also opened https://github.com/shinichi-takii/ddlparse/pull/62). Another benefit though is that my approach uses exactly the same config as for BigQuery, and then iterates over all instances, all databases and all tables within a given GCP project 😊

**Comment by rsyi** [comment]
Ah interesting. Surprising that those data types weren't already in there, especially since it seems `ddlparse` throws errors when those types are detected. 🤔  Looks like a straightforward PR to merge though!

FYI, I was just using the `informatinon_schema.columns` table, but the glaring problem is that type is always null for some reason:
https://cloud.google.com/spanner/docs/information-schema#information_schemacolumns

In any case, let me know if you need any help, and thanks for working on this!

**Comment by rsyi** [comment]
@manuelzander @biomunky 
Added type back in, and it all seems to work. There's a bit more information we could add in -- e.g. primary key (from the `constraints` table) -- but I'll leave this for now, as it'll require a bit of restructuring on the `TableMetadata` model.

Can you guys TAL and see if this fits with your workflow? 🙂  A couple things to note:
* I've used the following format for the filenames: `instance/database.schema.table`. It's a little inconsistent with the Bigquery naming pattern, which is `name/project_id.dataset.table`. **Thoughts on this, as it'll be quite permanent (if we change it, you'll have to run a migration later)?**
* Instances have to be specified (rather than looping through them all, like you had in your PR, @manuelzander). Is this acceptable, or do you have a ridiculous number of different spanner instances?

**Comment by rsyi** [comment]
@manuelzander @biomunky 
Added type back in, and it all seems to work. There's a bit more information we could add in -- e.g. primary key (from the `constraints` table) -- but I'll leave this for now, as it'll require a bit of restructuring on the `TableMetadata` model.

Can you guys TAL and see if this fits with your workflow? 🙂  A couple things to note:
* I've used the following format for the filenames: `instance/database.schema.table`. It's a little inconsistent with the Bigquery naming pattern, which is `name/project_id.dataset.table`. **Thoughts on this, as it'll be quite permanent (if we change it, you'll have to run a migration later)?**
* Instances have to be specified (rather than looping through them all, like you had in your PR, @manuelzander). Is this acceptable, or do you have a ridiculous number of different spanner instances?

**Comment by manuelzander** [comment]
@rsyi Personally, I think we should stay close to BigQuery naming pattern - my suggestion would be `name/project_id.instance.database.table` 😊 For me this seems the most natural way.
I'll push a commit with some changes reflecting this, let me know what you think.

**Comment by manuelzander** [comment]
LGTM, only the `pipelines/tests/unit/extractor/test_spanner_metadata_extractor.py` tests need to be adapted for the new approach.

File: pipelines/whale/extractor/spanner_metadata_extractor.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+                    columns.append(
+                        ColumnMetadata(
+                            row["col_name"],
+                            None,
+                            row["col_type"],
+                            row["col_sort_order"],
+                        )
+                    )
+
+                schema = "{}.{}".format(self._instance_id, self._database_id)
```
Comment (manuelzander) [inline pipelines/whale/extractor/spanner_metadata_extractor.py:133]: We could move this to line 118, doesn't need to be within the for loops 😊

Comment (rsyi) [reply pipelines/whale/extractor/spanner_metadata_extractor.py:133]: Good call.

**Pull request merged**